### ISSUE TITLE
[Question] odd ba* format selection?

### ISSUE THREAD

**Issue by probablykasper** [yt-dlp/yt-dlp#979]
## Checklist

- [x] I'm asking a question
- [x] I've looked through the README and FAQ for similar questions
- [x] I've searched the bugtracker for similar questions including closed ones

## Question

I'm running the following command:
```
yt-dlp -x -f 'ba*' --format-sort 'abr,codec' 'https://www.youtube.com/watch?v=nzLw6GMQ65I'
```

As a result, I get a 2.9MB `.m4a` file. When I pass `-F` it shows an ABR of `0k` (format ID 22).

If I instead use `-f 'ba'`, I get 3.3MB `.opus` file. Shouldn't `ba*` have selected this `.opus` file to begin with, since it's larger?

**Comment by pukkandan** [comment]
I can see why you would think that. But

> - `ba`, `bestaudio`: Select the best quality audio-only format. Equivalent to `best*[vcodec=none]`
> - `ba*`, `bestaudio*`: Select the best quality format that contains audio. It may also contain video. Equivalent to `best*[acodec!=none]`

**ba\*** is NOT "select the format that contains the **best quality audio**"

A format that contains a video stream is always considered better by yt-dlp than a format without it. Since 22 contains both an audio and a video stream, it is considered better than all audio-only streams. Hence why 22 gets selected. Hopefully I explained this clearly

---

But the more important question is **how do you select "the format that contains the best audio"**?

In theory, you can achieve this by asking yt-dlp to not consider video streams to be always better than audio-only (which is controlled by the sorting parameter `hasvid`) by using `--force-format-sort`. So `--force-format-sort -S "abr,acodec"` will give you what you want.

BUT, as you have already noticed, yt-dlp shows the `abr` of format 22 as `0k`. This is because we do not know the vbr and abr of this format, but only know the tbr. For simplicity, this tbr is shown in the vbr column when listing formats, with abr set to 0. This means we do not know what is the sound quality of this format and so will sort it with least priority when using the above method. So, this becomes the equivalent of just using `-S abr -f ba`

**tldr;** you can't for sure. Just use `ba` to get the best with the limited information we have

**Comment by probablykasper** [comment]
Ah I see, that explains it really well. Thanks!

**Comment by probablykasper** [comment]
Looks like the actual flag is `--format-sort-force`, not `-force-format-sort` like it mentions in the "Sorting Format" section in the docs

**Comment by pukkandan** [comment]
thanks I'll fix it
